src/expand_tilde.c

```diff
@@ -0,0 +1,140 @@
+ * ~: $HOME.
+ * ~+: $PWD.
+ * ~-: $OLDPWD.
+ * If not found increase update position to account for '~' and
+ * return ERR_NOT_FOUND.
+ * If found ft_strdup() the value.
+ * Calc replace.len.
+ *
+ * @param token Contains searched for token.
+ * @param symtab Environment.
+ * @param replace Where to save replace string.
+ * @param pos Current position.
+ * @return t_error SUCCESS, ERR_NOT_FOUND, ERR_MALLOC
+ */
+t_error	ft_get_tilde_replace(t_tok token, t_hashtable *symtab, t_tok *replace, size_t *pos)
+{
+	char		*target;
+	t_env_var	*env_var;
+
+	if (!ft_strcmp(token.str, "~"))
+		target = "HOME";
+	else if (!ft_strcmp(token.str, "~+"))
+		target = "PWD";
+	else if (!ft_strcmp(token.str, "~-"))
+		target = "OLDPWD";
+	env_var = ft_hashtable_lookup(symtab, target, ft_strlen(target));
+	if (!env_var)
+	{
+		(*pos)++;
+		return (ERR_NOT_FOUND);
+	}
+	replace->str = ft_strdup(env_var->value);
+	if (!replace->str)
+		return (ERR_MALLOC);
+	replace->len = ft_strlen(replace->str);
+	return (SUCCESS);
+}
+
+/**
+ * @brief Check for a recognised token.
+ *
+ * At first the string "~" is assigned.
+ * If '+' or '-' is found increase position and reassign
+ * Then a '/' or a '\0' has to be found, else the string is set to NULL.
+ * Also if not found increase the position to jump over the '~'.
+ * Calc the token.len.
+ *
+ * @param input String.
+ * @param pos Current position.
+ * @param token Where to save the token.
+ * @return t_error SUCCESS.
+ */
+t_error	ft_get_tilde_token(char *input, size_t *pos, t_tok *token)
+{
+	size_t	i;
+
+	i = *pos;
+	token->str = "~";
+	i++;
+	if (input[i] == '+')
+	{
+		token->str = "~+";
+		i++;
+	}
+	else if (input[i] == '-')
+	{
+		token->str = "~-";
+		i++;
+	}
+	if (input[i] != '/' && input[i] != '\0')
+	{
+		(*pos)++;
+		token->str = NULL;
+	}
+	else
+		token->len = ft_strlen(token->str);
+	return (SUCCESS);
+}
+
+/**
+ * @brief Expand a tilde char to corresponding environment var.
+ *
+ * If position is not 0 or the char before is not '=' return.
+ * Check if recognised token (~, ~+, ~-).
+ * Get the replacement for the found token.
+ * Insert the replacement in the string.
+ * Update the position to be after the inserted string.
+ *
+ * @param input String.
+ * @param symtab Environment.
+ * @param pos Current position.
+ * @return t_error SUCCESS, ERR_MALLOC.
+ */
+t_error	ft_expand_tilde(char **input, t_hashtable *symtab, size_t *pos)
+{
+	t_tok	token;
+	t_tok	replace;
+	t_error	err;
+
+	if (*pos != 0 && (*input)[*pos - 1] != '=')
+		return (SUCCESS);
+	token.str = NULL;
+	ft_get_tilde_token(*input, pos, &token);
+	if (!token.str)
+		return (SUCCESS);
+	replace.str = NULL;
+	err = ft_get_tilde_replace(token, symtab, &replace, pos);
+	if (err == ERR_NOT_FOUND)
+		return (SUCCESS);
+	if (err != SUCCESS)
+		return (err);
+	err = ft_insert_replace(input, *pos, token, replace);
+	free(replace.str);
+	if (err != SUCCESS)
+		return (err);
+	*pos += replace.len;
+	return (SUCCESS);
+}
```

Comment:
If a tilde is on the wrong position it doesn't get expanded. That's not an error per se: "hel~lo" is a valid string, it just means the tilde won't get expanded. 
Same case with no token found:  if you write ~. or ~* we don't (shouldn't) expand.
So it's still SUCCESS, even though no expansion happened.

Reply:
To my understanding a SUCCESS return only is justified, if the function has performed its role, in this case "expanded the tilde". If it has not done this, then I'd consider this an error in the sense that it could not do its job. Even though it might not be a fatal error causing the disruption of the whole program, I would call this sth like ERR_NOEXPAND.